Dump CS as package tree

rubenvereecken opened the pull request:
Changed dump to act more like [Il2CppDumper](https://github.com/Perfare/Il2CppDumper), dumping all classes as a file tree as opposed to one massive `.cs` file. Resulting a tree like

```
./UnityEngine.cs
./UnityEngine/UIModule.cs
./UnityEngine/PhysicsModule.cs
./UnityEngine/ScreenCaptureModule.cs
./UnityEngine/AssetBundleModule.cs
...
```

→ one file per assembly.

Currently only implemented for Android (making use of its `mkdir -p` like API), but happy to implement cross-platform if there is interest in merging this.

Also, if there's interest, happy to make this backward compatible by either adding a new function with current behaviour or an opt-in parameter.

rubenvereecken committed:
Implemented CS tree dump for Android

vfsfitvnm commented on src/dump.ts:
```diff
@@ -41,22 +41,35 @@ namespace Il2Cpp {
 
-        const destination = `${path ?? Il2Cpp.application.dataPath}/${fileName}`;
-        const file = new File(destination, "w");
+        const appContext = Java.use("android.app.ActivityThread").currentApplication().getApplicationContext()
```
Keep in mind we are multiplatorm here, but `Java`  is only available on Android

vfsfitvnm commented on src/dump.ts:
I would prefer to let the user choose whether to have a single big `.cs` file, or a directory tree. What do you think?

vfsfitvnm commented on src/dump.ts:
When creating a package tree, I think we should delete the tree first - otherwise we may end up mixing two applications (this cannot happen in mobile devices by the way...). It's not a hard requirement to me, but I want to hear your opinion

rubenvereecken replied:
Great point – I'm very much Android-centric. Maybe even a flag to allow deletion (because it's very destructive), but throw by default if anything exists there already.

vfsfitvnm commented:
Thanks!

> Currently only implemented for Android (making use of its mkdir -p like API), but happy to implement cross-platform if there is interest in merging this.

> Also, if there's interest, happy to make this backward compatible by either adding a new function with current behaviour or an opt-in parameter.

Yup, I left two comments to track these two changes 😄

rubenvereecken commented:
Great to see there's interest! I don't have an iOS device to test with, so I'll figure out how to have a cross-platform `mkdir -p` when I have a bit more time.

vfsfitvnm commented:
We have both 
```ts
const mkdir = new NativeFunction(Module.getExportByName(null, "mkdir"), "int", ["pointer", "int32"]);
```
and
```ts
Il2Cpp.corlib.class("System.IO").inner("Directory").method("CreateDirectory").invoke(Il2Cpp.string("..."));
```
:smile:

rubenvereecken committed:
mkdir -p based on libc to create directories recursively

rubenvereecken committed:
Detect if dump directory already exists

rubenvereecken commented:
Man, that's awesome, I need to get into the habit of using `corlib` more often!

I've now implemented `mkdir -p` using `libc.so` and moved it to a separate function for an extra clear API that users have to opt into.

That leaves us with `deleteIfExists`. I've covered the case of `false`, which just errors, then exits. For tree deletion I need a `rm -rf`, which is [non-trivial to implement using libc](https://stackoverflow.com/questions/5467725/how-to-delete-a-directory-and-its-contents-in-posix-c), so I'll need a minute for that.

Finally, it's not super-optimised in that there could probably be less `mkdir` calls and we could cache the function lookups, but I figured it's a one-off API anyway.

---

As an aside, 
```
Il2Cpp.corlib.class("System.IO").inner("Directory")
```
wasn't available for me – I needed 
```
Il2Cpp.corlib.class("System.IO.Directory")
```
Granted, on Unity 2022.x, and I was too lazy to compile on 2020 or something. Although I just realised `inner` isn't actually in the API, so maybe that was just you writing off the cuff. 

Edit: I realise you probably meant `class.nested(...)`.

rubenvereecken committed:
Simplified dumpTree using mscorlib System.IO.Directory

rubenvereecken commented:
@vfsfitvnm alright, I simplified it considerably using `System.IO.Directory`. However, my corlib doesn't seem to come with a [recursive delete](https://learn.microsoft.com/en-us/dotnet/api/system.io.directory.delete?view=netframework-4.0#system-io-directory-delete(system-string-system-boolean)), so I decided to not over-engineer this and simply warn + exit if the directory already exists, so as not to accidentally merge dumps. 

Good for final review + merge I reckon?

vfsfitvnm commented:
> Edit: I realise you probably meant class.nested(...).

Exactly 😄 It's been a while since I used my own module ahah.

> Granted, on Unity 2022.x, and I was too lazy to compile on 2020 or something.

This project has `test`/`unity` folder that serves this purpose, however there is no documentation so yeah I didn't expect you to guess how it works!

> Good for final review + merge I reckon?

Let's see!

vfsfitvnm commented on src/dump.ts:
Let's use the double quotes (`"`) instead of the single quotes (`'`) to wrap string literals

vfsfitvnm commented on src/dump.ts:
```diff
@@ -59,4 +63,64 @@ namespace Il2Cpp {
+        const Directory = Il2Cpp.corlib.class('System.IO.Directory');
+        const CreateDirectory = Directory.method('CreateDirectory');
+        CreateDirectory.invoke(Il2Cpp.string(p));
+    }
```
Simple expressions can be inlined (please apply to `directoryExists` and `removeDirectoryTree` as well)
```suggestion
    function createDirectoryRecursively(path: string) {
        Il2Cpp.corlib.class("System.IO.Directory").method("CreateDirectory").invoke(Il2Cpp.string(path));
    }
```

rubenvereecken replied:
I'll just keep `createDirectoryRecursively` (inlined its contents) because it's used in two places

vfsfitvnm commented on src/dump.ts:
```diff
@@ -59,4 +63,64 @@ namespace Il2Cpp {
+            // Create directory (recursively) if necessary
+            createDirectoryRecursively(path);
+
+            const destination = `${path}/${filename}`;    
```
Is it equivalent?

```suggestion
        for (const assembly of Il2Cpp.domain.assemblies) {
            const destination = assembly.name.replaceAll(".", "/") + ".cs";

            createDirectoryRecursively(destination.substring(0, destination.lastIndexOf("/")));
```

rubenvereecken replied:
Almost, still needs to include the `basePath`, but I'll simplify a bit along the lines of your flavour.

vfsfitvnm commented on src/dump.ts:
```diff
@@ -59,4 +63,64 @@ namespace Il2Cpp {
+            const destination = `${path}/${filename}`;    
+            const file = new File(destination, "w");
+
+            inform(`dumping ${path}/${filename}`);
```
I would keep as it is in `dump`:
```suggestion
            inform(`dumping ${assembly.name}...`);
```

vfsfitvnm replied:
I would move this log as the first instruction within the `for` loop

rubenvereecken committed:
Processed review and cleaned up

vfsfitvnm commented on src/dump.ts:
```diff
@@ -59,4 +63,44 @@ namespace Il2Cpp {
+        }
+
+        for (const assembly of Il2Cpp.domain.assemblies) {
+            const assemblyPath = assembly.name.replace(".", "/") + ".cs";
```
```suggestion
            const assemblyPath = assembly.name.replaceAll(".", "/") + ".cs";
```

rubenvereecken replied:
Odd, yeah, thought I'd tested this. I'll need to bump `lib` to `es2021` in tsconfig.

vfsfitvnm commented on src/dump.ts:
```diff
@@ -59,4 +63,44 @@ namespace Il2Cpp {
+    }
+
+    function createDirectoryRecursively(p: string) {
+        return Il2Cpp.corlib.class("System.IO.Directory").method("CreateDirectory").invoke(Il2Cpp.string(p));
```
```suggestion
    function createDirectoryRecursively(path: string) {
        return Il2Cpp.corlib.class("System.IO.Directory").method("CreateDirectory").invoke(Il2Cpp.string(path));
```

vfsfitvnm commented on src/dump.ts:
```diff
@@ -59,4 +63,44 @@ namespace Il2Cpp {
+        if (deleteIfExists && basePathExists) {
+            warn(`directory ${basePath} already exists, but tree deletion not yet supported, skipping...`);
+            return;
+        }
```
If deleting a directory is not supported, I would remove all related code:
```suggestion
    export function dumpTree(path?: string): void {
        const basePath = path ?? `${Il2Cpp.application.dataPath!}/${Il2Cpp.application.identifier ?? "unknown"}_${Il2Cpp.application.version ?? "unknown"}`;
        const basePathExists = Il2Cpp.corlib.class("System.IO.Directory").method<boolean>("Exists").invoke(Il2Cpp.string(basePath));

        if (basePathExists) {
            warn(`directory ${basePath} already exists, skipping...`);
            return;
        }
```

rubenvereecken committed:
Update src/dump.ts

Co-authored-by: vfsfitvnm <46219656+vfsfitvnm@users.noreply.github.com>

rubenvereecken committed:
Update src/dump.ts

Co-authored-by: vfsfitvnm <46219656+vfsfitvnm@users.noreply.github.com>

rubenvereecken committed:
Bump lib to es2021 so we can have String.replaceAll

rubenvereecken commented on tsconfig.json:
```diff
@@ -1,18 +1,16 @@
-  }
+    "compilerOptions": {
+        "target": "es2022",
+        "lib": ["es2021"],
```
All tests run fine with this new `lib` setting.

vfsfitvnm replied:
Thanks, I see!

vfsfitvnm commented on src/dump.ts:
```diff
@@ -59,4 +63,39 @@ namespace Il2Cpp {
+            inform(`dumping ${assemblyPath}`);
+
+            // Create directory (recursively) if necessary
+            createDirectoryRecursively(destination.substring(0, destination.lastIndexOf("/")));
```
Style change
```suggestion
            inform(`dumping ${assembly.name}...`);

            const destination = `${basePath}/${assembly.name.replaceAll(".", "/")}.cs`;

            createDirectoryRecursively(destination.substring(0, destination.lastIndexOf("/")));
```

rubenvereecken committed:
Update src/dump.ts

Co-authored-by: vfsfitvnm <46219656+vfsfitvnm@users.noreply.github.com>

vfsfitvnm commented on src/dump.ts:
```diff
@@ -59,4 +63,37 @@ namespace Il2Cpp {
+        if (basePathExists) {
+            warn(`directory ${basePath} already exists, skipping...`);
+            return;
+        }
```
I would add a parameter (e.g. `ignoreAlreadyExistingDirectory: boolean = false`) to allow writing to an already existing folder. Please extract `directoryExists` to a function (just like `createDirectoryRecursively`):

```suggestion
    export function dumpTree(path?: string, ignoreAlreadyExistingDirectory: boolean = false): void {
        const basePath = path ?? `${Il2Cpp.application.dataPath!}/${Il2Cpp.application.identifier ?? "unknown"}_${Il2Cpp.application.version ?? "unknown"}`;

        if (!ignoreAlreadyExistingDirectory && directoryExists(basePath)) {
            raise(`directory ${basePath} already exists - pass ignoreAlreadyExistingDirectory = true to skip this check`);
        }
```

rubenvereecken replied:
Ah, not _that_ kind of inlining 👍

rubenvereecken committed:
Added directoryExists back in

rubenvereecken committed:
Update src/dump.ts

Co-authored-by: vfsfitvnm <46219656+vfsfitvnm@users.noreply.github.com>

vfsfitvnm commented on src/dump.ts:
```diff
@@ -59,4 +63,39 @@ namespace Il2Cpp {
+
+    function directoryExists(path: string) {
+        Il2Cpp.corlib.class("System.IO.Directory").method<boolean>("Exists").invoke(Il2Cpp.string(path));
+    }
```
```suggestion
    function directoryExists(path: string): boolean {
        return Il2Cpp.corlib.class("System.IO.Directory").method<boolean>("Exists").invoke(Il2Cpp.string(path));
    }
```

rubenvereecken committed:
Update src/dump.ts

Co-authored-by: vfsfitvnm <46219656+vfsfitvnm@users.noreply.github.com>

vfsfitvnm committed:
Fix formatting of tsconfig.json

vfsfitvnm committed:
Remove comment

vfsfitvnm committed:
Reuse 'path' variable

vfsfitvnm committed:
Add doc comment

vfsfitvnm commented:
I pushed a few commits to polish the code.

Note to self: if other dump modes are required in the future, consider this rusty TypeScript code:
```ts
    export function dump(mode: "file" | "tree", options: DumpOptions<typeof mode>) {
        // ...
    }

    type DumpOptions<Mode extends "file" | "tree"> = Mode extends "file"
        ? {
              path?: string;
              fileName?: string;
          }
        : Mode extends "tree"
        ? {
              path?: string;
              ignoreAlreadyExistingDirectory?: boolean;
          }
        : never;
```

vfsfitvnm committed:
Remove unused return value in createDirectoryRecursively

rubenvereecken commented:
Well that was easy 😄

vfsfitvnm commented:
Merged! Thanks :confetti_ball:

rubenvereecken commented:
Awesome
